File: Python/005__Longest_Palindromic_Substring.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: str
        """
        def count(s, l, r):
            while l >= 0 and r < len(s) and s[l] == s[r]:
                l -= 1
                r += 1
            return r - l - 1


        if len(s) == 0:
            return ""
        start, end = 0, 0
        for i in range(len(s)):
            len1 = count(s, i, i)
            len2 = count(s, i, i + 1)
            length = max(len1, len2)
            if length > end - start:
                start = i - (length - 1) / 2
                end = i + length / 2
        return s[start : end + 1]
        

File: Java/005__Longest_Palindromic_Substring.java
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) return "";
        int start = -1, end = -1;
        for (int i = 0; i < s.length(); i++) {
            int len1 = helper(s, i, i);
            int len2 = helper(s, i, i + 1);
            int len = Math.max(len1, len2);
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        return s.substring(start, end + 1);
    }

    private int helper(String s, int l, int r) {
        int L = l, R = r;
        while (L >= 0 && R < s.length() && s.charAt(L) == s.charAt(R)) {
            L--;
            R++;
        }
        return R - L - 1;
    }
}
